### PR TITLE
Fix `Item slots do not start hovered on formspec creation/update if the cursor is on them`

### PR DESCRIPTION
**Pull request by Mahoyomu**
- Goal of the PR
- solve [`Item slots do not start hovered on formspec creation/update if the cursor is on them`](https://github.com/minetest/minetest/issues/14530)
- How does the PR work?
- By keeping the hovering status during the formspec update, and keeping tooltip box on.
- Does it resolve any reported issue?
- [`Item slots do not start hovered on formspec creation/update if the cursor is on them`](https://github.com/minetest/minetest/issues/14530)
- Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?
- IMHO yes, for many mods can use similar formspec update logic like furnace, which will likely trigger this bug. It's not a expected behavior and maybe kind of annoying to users in [Wuzzy2](https://github.com/Wuzzy2)'s words.
- If not a bug fix, why is this PR needed? What usecases does it solve?
- This is a bug fix.

## To do

Keep the hovering status during the formspec update, and keep tooltip box on

This PR is Ready for Review.

## How to test
Compile and run, or see changes in the src.

### CODE REVIEW

**Comment by Wuzzy2** [comment]
@Mahoyomu: Yes.

**Review by appgurueu — COMMENTED** [review]
Thanks for the bugfix!

I can reproduce the bug using a MTG furnace as Wuzzy described. With this PR, it seems to work as expected.

However, I don't see the purpose of `is_last_mouse_event_outside_item_slot`. Why can't you omit it and always recalculate `m_hovered_i` (when it is `-1`) in the draw method?

The bug stems from the fact that when a new formspec is received, the GUI is regenerated, and a new GUI inventory list is created, where `m_hovered_i` is wrongly initialized to `-1`. This is only updated in `OnEvent`, causing the bug.

It seems to me the obvious (and somewhat cleaner) solution is to initialize `m_hovered_i` properly in the `GUIInventoryList` constructor instead of updating it in the draw method. The following patch also seems to work:

```diff
diff --git a/src/gui/guiInventoryList.cpp b/src/gui/guiInventoryList.cpp
index 02505d436..991cd9b74 100644
--- a/src/gui/guiInventoryList.cpp
+++ b/src/gui/guiInventoryList.cpp
@@ -21,6 +21,8 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "guiFormSpecMenu.h"
 #include "client/hud.h"
 #include "client/client.h"
+#include "client/renderingengine.h"
+#include "ICursorControl.h"
 
 GUIInventoryList::GUIInventoryList(gui::IGUIEnvironment *env,
 	gui::IGUIElement *parent,
@@ -47,9 +49,10 @@ GUIInventoryList::GUIInventoryList(gui::IGUIEnvironment *env,
 	m_fs_menu(fs_menu),
 	m_options(options),
 	m_font(font),
-	m_hovered_i(-1),
 	m_already_warned(false)
 {
+	ICursorControl *cursor_control = RenderingEngine::get_raw_device()->getCursorControl();
+	m_hovered_i = getItemIndexAtPos(v2s32(cursor_control->getPosition()));
 }
 
 void GUIInventoryList::draw()
```

**Comment by appgurueu** [comment]
> I have to apologize that I don't really understand this, please help me with a commit if you could.

Sure. I have switched this PR to initializing the variable properly in the constructor instead.

**Comment by Desour** [comment]
I see two ways how this won't work correctly:
* If another (non point-through-able) element (e.g. a button) is above the invlist slot, it has to block the hovering. (Checking this in `GUIInventoryList` will be hard and ugly.)
* When the formspec opens, it's possible that the invlist is not yet sent to the player (see also  #9640). `getItemIndexAtPos` doesn't work in this case.
  (This should be easy to fix by not taking the invlist size into account in `getItemIndexAtPos`, and adjusting other callers (i.e. `GUIFormSpecMenu::getItemAtPos`) to not be broken.)

**Comment by appgurueu** [comment]
> I think that's another issue now

Agreed on the first point - fixing `getItemIndexAtPos` to not rely on the inventory list needn't be part of this PR - but the second point is related to this PR. For this PR to correctly mark slots as hovered, it would need to take this into account.

This problem ultimately stems from sidestepping the event system. What if instead we somehow used the event system, e.g. by emitting a fake mouse move event (or perhaps cleaner, a GUI event or something similar) when a formspec is regenerated?

File: src/gui/guiInventoryList.cpp [hunk 1]
```diff
@@ -47,9 +49,11 @@ GUIInventoryList::GUIInventoryList(gui::IGUIEnvironment *env,
 	m_fs_menu(fs_menu),
 	m_options(options),
 	m_font(font),
-	m_hovered_i(-1),
 	m_already_warned(false)
 {
+	// FIX ME
```
Comment (sfan5) [inline src/gui/guiInventoryList.cpp:54]: what is there to fix? please describe it.

**Comment by Desour** [comment]
> > * e invlist size
> 
> You mean like delete `(s32)ilist->getSize()` in `if (rect.getArea() > 0 && rect.isPointInside(p) && i + m_start_item_i < (s32)ilist->getSize())` , and move this judgement to other callers ?

Yes.

(Weird quote though.)

**Review by sfan5 — APPROVED** [review]
looks correct

**Comment by appgurueu** [comment]
> I'm sorry but I really don't understand how to rebase it

Essentially you run `git rebase master` and then you resolve conflicts. For how to do that, see tutorials such as https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase. See also https://dev.minetest.net/Git.

> I only know about merging

Resolving conflicts by merging master into into this branch (as it seems you have) is in this case acceptable since this PR will just be squashed at the end.

**Comment by sfan5** [comment]
Sorry, but this needs a rebase (or merge) again.

**Pull request closed without merging**